# **Advanced Pandas Assignment**

# **Assignment Questions**

# **Load The Dataset**

In [23]:
import pandas as pd

df = pd.read_csv("Advanced_pandas_sales_dataset.csv")

df.head()

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30


## **Q1 Dataset Exploration**
Perform the following exploratory steps:  
- Display the first 10 rows  
- Display the dataset structure using info()  
- Find the number of rows and columns  
- Identify the data types of each column  

## **Answer:**

## **First 10 rows**

In [4]:
df.head(10)

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
0,1001,2024-10-01,Priya,Hyderabad,East,Mouse,Accessories,1735,7,0.00,307.52
1,1002,2024-05-03,Sneha,Kolkata,East,Monitor,Electronics,807,2,0.15,49.89
2,1003,2024-01-26,Anita,Mumbai,West,Phone,Electronics,2328,1,0.20,66.47
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
6,1007,2024-06-24,Dev,Chennai,East,Monitor,Electronics,1170,5,0.20,22.83
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
9,1010,2024-12-25,Pooja,Ahmedabad,North,Phone,Electronics,2711,8,0.00,477.32


## **Dataset Structure**

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB


## **No of rows and columns**

In [6]:
df.shape

(230, 11)

## **Data types of each column**

In [7]:
df.dtypes

Order_ID        int64
Order_Date     object
Customer       object
City           object
Region         object
Product        object
Category       object
Sales           int64
Quantity        int64
Discount      float64
Profit        float64
dtype: object

## **Q2 Sales Filtering Analysis**  
Filter the dataset to display:  
1. Orders where Sales > 3500  
2. Orders where Region = "East" and Sales > 2500  
Explain what insights you observe.  

## **Answer:**

## **1. Sales > 3500**

In [8]:
df[df['Sales'] > 3500]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
10,1011,2024-06-30,Neha,Delhi,South,Printer,Office,5977,6,0.00,1578.32
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


## **2. Region = "East" and Sales > 2500**

In [9]:
df[(df['Region'] == 'East') & (df['Sales'] > 2500)]

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
13,1014,2024-03-18,Anita,Bangalore,East,Mouse,Accessories,3616,4,0.05,463.93
21,1022,2024-01-14,Nitin,Chennai,East,Monitor,Electronics,4827,6,0.05,912.34
27,1028,2024-07-21,Dev,Kolkata,East,Tablet,Electronics,4054,6,0.00,1013.83
33,1034,2024-12-15,Nitin,Ahmedabad,East,Tablet,Electronics,4378,2,0.20,-104.30
34,1035,2024-02-20,Nitin,Ahmedabad,East,Laptop,Electronics,2535,4,0.10,188.15
41,1042,2024-07-15,Rohan,Chennai,East,Laptop,Electronics,3268,8,0.15,295.15
43,1044,2024-09-20,Kavya,Hyderabad,East,Printer,Office,2618,7,0.05,454.64


## **Insights I have observe**

- Orders like **3933, 4378 and 4811 sales** show high-value transactions.  
- Some high sales still give **negative profit (e.g., -71)** → cost/discount issue.  
- East region contains moderate-high sales but not always profitable.

## **Q3 Customer Purchase Analysis**  
Identify the top 5 customers based on total sales.  
Hint: Use GroupBy with aggregation.

## **Answer:**

## **Top 5 Customers**

In [10]:
df.groupby('Customer')['Sales'].sum().sort_values(ascending=False).head(5)

Customer
Nitin    59820
Raj      58416
Rahul    52291
Dev      48500
Isha     46818
Name: Sales, dtype: int64

## **Q4 Regional Sales Performance**  
Calculate the total sales generated by each region.  
Then identify:  
- Which region generated the highest revenue?

## **Answer:**

## **Total Sales by each Region**

In [13]:
region_sales = df.groupby('Region')['Sales'].sum()
region_sales

Region
East     232675
North    158210
South    183108
West     170227
Name: Sales, dtype: int64

## **Region generated the highest revenue**

In [14]:
region_sales.idxmax()

'East'

## **Q5 Product Performance Analysis**  
- Calculate the total quantity sold for each product.
- Which product appears to be the most popular?

## **Answer:**

## **Total quantity sold for each product**

In [15]:
product_quantity = df.groupby('Product')['Quantity'].sum()
product_quantity

Product
Headphones    115
Keyboard      110
Laptop        140
Monitor       156
Mouse         122
Phone         151
Printer       108
Tablet        150
Name: Quantity, dtype: int64

## **Product appears to be the most popular**

In [16]:
product_quantity.idxmax()

'Monitor'

## **Q6 Multi-Level Grouping**  
- Group the dataset by: Region and Category
- Calculate the total sales for each combination.
- What insights can you derive from this analysis?

## **Answer:**

## **Group by Region & Category (Total Sales)**

In [17]:
df.groupby(['Region', 'Category'])['Sales'].sum()

Region  Category   
East    Accessories     57646
        Electronics    160907
        Office          14122
North   Accessories     62052
        Electronics     70023
        Office          26135
South   Accessories     80914
        Electronics     86690
        Office          15504
West    Accessories     64811
        Electronics     84128
        Office          21288
Name: Sales, dtype: int64

## **Insight**
- Shows which category performs best in each region
- Electronics dominate in **All** regions.

## **Q7 Pivot Table Creation**
Create a pivot table showing:
- Rows → Region
- Columns → Product
- Values → Total Sales
Which region sells the most laptops?

## **Answer:**

## **Pivot Table**

In [18]:
pivot = pd.pivot_table(df, values='Sales', index='Region', columns='Product', aggfunc='sum')
pivot

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,23899,10845,55568,31299,22902,39667,14122,34373
North,15450,20538,4540,14922,26064,30535,26135,20026
South,24111,23480,23029,19167,33323,18218,15504,26276
West,17159,38499,10826,18178,9153,27850,21288,27274


## **Region Selling Most Laptop**

In [19]:
pivot['Laptop'].idxmax()

'East'

## **Q8 Profitability Analysis**
- Calculate the average profit for each product category.
- Which category is most profitable?

## **Answer:**

## **average profit for each product category**

In [20]:
df.groupby('Category')['Profit'].mean()

Category
Accessories    410.603086
Electronics    396.462114
Office         471.938462
Name: Profit, dtype: float64

## **Most Profitable Category**

In [21]:
df.groupby('Category')['Profit'].mean().idxmax()

'Office'

## **Q9 Discount Impact Study**
Analyze the relationship between discount and profit.  
Tasks:  
- Calculate the average profit for each discount level
- Identify whether higher discounts lead to lower profit

## **Answer:**

## **Average Profit for each discount level**

In [25]:
df.groupby('Discount')['Profit'].mean()

Discount
0.00    751.210588
0.05    531.368039
0.10    346.460513
0.15    237.191400
0.20     90.026923
Name: Profit, dtype: float64

## **Insight**

Example already shows:  
Discount 0.20 → negative profit case exists  
Conclusion:  
Higher discount → lower or negative profit (clear trend)

## **Q10 End-to-End Data Wrangling Task**
Perform the following workflow:
1. Filter the dataset where Sales > 3000
2. Group data by Region and Product
3. Calculate total sales
4. Reshape the result into a pivot table  
Explain the business insights obtained from this analysis.

## **Answer:**

## **1. Filter Dataset where Sales > 3000**

In [26]:
filtered_df = df[df['Sales'] > 3000]
filtered_df

,Order_ID,Order_Date,Customer,City,Region,Product,Category,Sales,Quantity,Discount,Profit
3,1004,2024-06-06,Sara,Pune,East,Keyboard,Accessories,3933,3,0.20,-71.66
4,1005,2024-10-15,Rahul,Bangalore,South,Monitor,Electronics,3550,2,0.20,146.30
5,1006,2024-07-04,Pooja,Bangalore,West,Printer,Office,3073,8,0.20,191.58
7,1008,2024-03-18,Kavya,Kolkata,East,Phone,Electronics,4693,7,0.05,918.49
8,1009,2024-10-31,Kavya,Ahmedabad,East,Phone,Electronics,5071,6,0.10,707.55
...,...,...,...,...,...,...,...,...,...,...,...
219,1220,2024-03-22,Meera,Bangalore,West,Phone,Electronics,5986,7,0.05,986.70
222,1223,2024-03-04,Meera,Ahmedabad,South,Phone,Electronics,4356,6,0.10,416.87
223,1224,2024-01-07,Dev,Bangalore,East,Tablet,Electronics,4115,4,0.05,870.03
225,1226,2024-06-14,Anita,Ahmedabad,East,Laptop,Electronics,5647,2,0.15,662.00


## **2. Group Data by Region and Product**

In [27]:
grouped_data = filtered_df.groupby(['Region', 'Product'])['Sales'].sum()
grouped_data

Region  Product   
East    Headphones    14469
        Keyboard       9642
        Laptop        44096
        Monitor       16420
        Mouse         19000
        Phone         29500
        Printer        3696
        Tablet        29229
North   Headphones     5326
        Keyboard      15653
        Laptop         4540
        Monitor        5087
        Mouse         22209
        Phone         24670
        Printer       17185
        Tablet         3392
South   Headphones    10026
        Keyboard      19249
        Laptop        17950
        Monitor       14921
        Mouse         26928
        Phone         18218
        Printer       14143
        Tablet        19613
West    Headphones    13898
        Keyboard      33380
        Laptop         5922
        Monitor       14354
        Mouse          3511
        Phone         15653
        Printer       12648
        Tablet        17185
Name: Sales, dtype: int64

## **3. Calculate Total Sales**

In [28]:
total_sales = df['Sales'].sum()
total_sales

np.int64(744220)

## **4. Reshape Data into Pivot Table**

In [29]:
pivot_table = grouped_data.unstack()
pivot_table

Product,Headphones,Keyboard,Laptop,Monitor,Mouse,Phone,Printer,Tablet
Region,,,,,,,,
East,14469,9642,44096,16420,19000,29500,3696,29229
North,5326,15653,4540,5087,22209,24670,17185,3392
South,10026,19249,17950,14921,26928,18218,14143,19613
West,13898,33380,5922,14354,3511,15653,12648,17185


## **Business Insight**

- High-value transactions (Sales > 3000) are limited but contribute significantly to revenue.
- Sales are concentrated in specific Region–Product combinations, indicating targeted demand patterns.
- Certain regions dominate in particular products, suggesting localized customer preferences.
- High sales do not always guarantee high profit, indicating the impact of discounts and costs.
- This analysis helps businesses focus on high-performing regions and products for better revenue optimization and strategic decision-making.
- The pivot table enables quick comparison and supports data-driven decisions in marketing, inventory, and regional expansion strategies.